File: hyperscalp-ai-main/base44/functions/botExecutor/strategies/trendBreakout.ts
/**
 * Strategy B — Trend Continuation Breakout.
 *
 * Thesis: in an established HTF trend (1h EMA stack + 15m ADX>25), pullbacks
 * resolve in the trend's direction. Enter on a 1m close that breaks the prior
 * 5m high (or low for shorts), but ONLY when CVD also makes a new extreme
 * (real divergence check, not slope). This filters out fake breakouts where
 * price ticks higher but aggressor flow doesn't follow.
 *
 * Exit: partial TP at 1R, chandelier trail on the runner.
 */
import type { StrategyContext, StrategyResult } from './types.ts';
import { atr, cvdSeries } from '../lib/features.ts';

export function trendBreakout(ctx: StrategyContext): StrategyResult {
  const cfg = ctx.cfg;
  if (!cfg?.enabled) return { signal: null, reject: 'disabled' };
  if (ctx.regime !== 'TREND_UP' && ctx.regime !== 'TREND_DOWN') return { signal: null, reject: `regime=${ctx.regime}` };
  if (ctx.c1m.length < 60) return { signal: null, reject: 'insufficient 1m candles' };

  const direction: 'long' | 'short' = ctx.regime === 'TREND_UP' ? 'long' : 'short';

  const lookback = cfg.breakout_lookback_min ?? 5;
  const recent = ctx.c1m.slice(-(lookback + 1), -1);
  if (recent.length < lookback) return { signal: null, reject: 'breakout window too small' };
  const recentHigh = Math.max(...recent.map(c => c.h));
  const recentLow = Math.min(...recent.map(c => c.l));

  const last = ctx.c1m[ctx.c1m.length - 1];
  const price = last.c;

  // Breakout check on the just-closed bar
  const broke = direction === 'long' ? last.c > recentHigh : last.c < recentLow;
  if (!broke) return { signal: null, reject: `no breakout (price=${price} vs ${direction === 'long' ? recentHigh : recentLow})` };

  // CVD must also make a new extreme over the same window — this is the real divergence check
  if (cfg.require_cvd_new_extreme) {
    const { cvd } = cvdSeries(ctx.tape, (lookback + 2) * 60_000);
    if (cvd.length < 4) return { signal: null, reject: 'insufficient CVD samples' };
    const lastCvd = cvd[cvd.length - 1];
    const priorCvd = cvd.slice(0, -1);
    const cvdNewExtreme = direction === 'long'
      ? lastCvd >= Math.max(...priorCvd)
      : lastCvd <= Math.min(...priorCvd);
    if (!cvdNewExtreme) return { signal: null, reject: 'price broke but CVD did not (fake breakout)' };
  }

  // BTC leadership: alts shouldn't break out alone if BTC isn't moving with us
  if (ctx.coin !== 'BTC') {
    if (direction === 'long' && ctx.btcLead < 0) return { signal: null, reject: 'BTC dumping; alt long unsafe' };
    if (direction === 'short' && ctx.btcLead > 0) return { signal: null, reject: 'BTC pumping; alt short unsafe' };
  }

  const a1 = atr(ctx.c1m, 14);
  // Stop just outside the breakout structure (other side of recent range)
  const stop = direction === 'long' ? recentLow - a1 * 0.2 : recentHigh + a1 * 0.2;
  const r = Math.abs(price - stop);
  // Initial target for cost-filter purposes is 1R (partial TP). The runner trails.
  const target = direction === 'long' ? price + r * 1.0 : price - r * 1.0;

  return {
    signal: {
      strategy: 'trend_breakout',
      direction,
      entry: price,
      stop, target,
      expectedHoldingMinutes: 18,
      preferredEntryMode: 'depth_aware',
      partialTpR: cfg.partial_tp_r ?? 1.0,
      partialTpFraction: cfg.partial_tp_fraction ?? 0.5,
      trailType: 'chandelier',
      trailAtrMult: cfg.chandelier_atr_mult ?? 2.5,
      reason: `TREND ${direction.toUpperCase()} | brk=${(direction === 'long' ? recentHigh : recentLow).toFixed(4)} | adx=${'computed in regime'}`,
      features: {
        breakout_level: direction === 'long' ? recentHigh : recentLow,
        atr: a1, r_distance: r,
      },
    },
  };
}

File: hyperscalp-ai-main/research/backtester/strategies/trend_breakout.py
"""Python port of strategies/trendBreakout.ts."""
from dataclasses import dataclass
from typing import Optional
import pandas as pd

from ..features import atr, adx, ema_stack, cvd_series


@dataclass
class Signal:
    strategy: str; direction: str
    entry: float; stop: float; target: float
    expected_holding_minutes: int
    preferred_entry_mode: str
    partial_tp_r: float; partial_tp_fraction: float
    trail_type: str; trail_atr_mult: float
    reason: str


def trend_breakout_signal(bars1m: pd.DataFrame, bars15m: pd.DataFrame, bars1h: pd.DataFrame, trades: pd.DataFrame, btc_lead: int, cfg: dict) -> Optional[Signal]:
    if len(bars1m) < 60 or len(bars15m) < 30 or len(bars1h) < 50:
        return None
    a15 = adx(bars15m, 14)
    if a15 < cfg.get("min_adx_15m", 25):
        return None
    htf = ema_stack(bars1h)
    if cfg.get("ema_stack_required", True) and htf == 0:
        return None
    direction = "long" if htf == 1 else "short" if htf == -1 else None
    if direction is None:
        return None

    lb = cfg.get("breakout_lookback_min", 5)
    recent = bars1m.iloc[-(lb + 1):-1]
    if len(recent) < lb:
        return None
    recent_high = float(recent["high"].max())
    recent_low = float(recent["low"].min())
    last = bars1m.iloc[-1]
    price = float(last["close"])

    if direction == "long" and not (price > recent_high):
        return None
    if direction == "short" and not (price < recent_low):
        return None

    if cfg.get("require_cvd_new_extreme", True) and not trades.empty:
        end_ts = bars1m.index[-1]
        cvd, _ = cvd_series(trades, end_ts, (lb + 2) * 60_000)
        if len(cvd) >= 4:
            last_cvd = cvd[-1]
            prior = cvd[:-1]
            if direction == "long" and last_cvd < max(prior):
                return None
            if direction == "short" and last_cvd > min(prior):
                return None

    a1 = atr(bars1m, 14)
    stop = recent_low - a1 * 0.2 if direction == "long" else recent_high + a1 * 0.2
    r = abs(price - stop)
    target = price + r * 1.0 if direction == "long" else price - r * 1.0

    return Signal(
        strategy="trend_breakout", direction=direction, entry=price, stop=stop, target=target,
        expected_holding_minutes=18, preferred_entry_mode="depth_aware",
        partial_tp_r=cfg.get("partial_tp_r", 1.0), partial_tp_fraction=cfg.get("partial_tp_fraction", 0.5),
        trail_type="chandelier", trail_atr_mult=cfg.get("chandelier_atr_mult", 2.5),
        reason=f"TREND {direction.upper()} brk={recent_high if direction == 'long' else recent_low:.4f} adx15={a15:.1f}",
    )
